[Intelligence Effects] Support adding intelligence effects to Writing Tools Proofreading operations

rr-codes opened the pull request:
#### fc0f18354ef6bd05c8504089231bb20f50e6b685
<pre>
[Intelligence Effects] Support adding intelligence effects to Writing Tools Proofreading operations
<a href="https://bugs.webkit.org/show_bug.cgi?id=282161">https://bugs.webkit.org/show_bug.cgi?id=282161</a>
<a href="https://rdar.apple.com/138751347">rdar://138751347</a>

Reviewed by Aditya Keerthi.

When performing Proofreading, the text will now animate using the intelligence text effects.

The implementation details are described throughout the change in comments.

* Source/WebCore/page/IntelligenceTextEffectsSupport.cpp:
(WebCore::IntelligenceTextEffectsSupport::updateTextVisibility):
(WebCore::IntelligenceTextEffectsSupport::decorateWritingToolsTextReplacements):
* Source/WebCore/page/IntelligenceTextEffectsSupport.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::proofreadingSessionDidReceiveSuggestions):
(WebCore::Page::updateTextVisibilityForActiveWritingToolsSession):
(WebCore::Page::decorateTextReplacementsForActiveWritingToolsSession):
(WebCore::Page::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::didBeginWritingToolsSession):
(WebCore::WritingToolsController::proofreadingSessionDidReceiveSuggestions):
(WebCore::WritingToolsController::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/ObjectiveCBlockConversions.swift:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView proofreadingSession:didReceiveSuggestions:processedRange:inContext:finished:]):
(-[WKWebView didEndWritingToolsSession:accepted:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:identifier:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:decorateReplacementsForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:completion:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::proofreadingSessionDidReceiveSuggestions):
(WebKit::WebPageProxy::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPageProxy::decorateTextReplacementsForActiveWritingToolsSession):
(WebKit::WebPageProxy::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:
(PlatformIntelligenceTextEffectViewSource.replacementEffectDidComplete(_:)):
(UITextEffectViewSourceAdapter.targetedPreview(for:)):
(UITextEffectViewSourceAdapter.updateTextChunkVisibilityForAnimation(_:visible:)):
(UIReplacementTextEffectDelegateAdapter.replacementEffectDidComplete(_:)):
(UIReplacementTextEffectDelegateAdapter.performReplacementAndGeneratePreview(for:effect:animation:)):
(WTTextPreviewAsyncSourceAdapter.textPreviews(for:)):
(WTTextPreviewAsyncSourceAdapter.updateIsTextVisible(_:for:)):
(PlatformIntelligenceTextEffectView.wrappedEffectIDToPlatformEffects):
(PlatformIntelligenceTextEffectView.frame):
(PlatformIntelligenceTextEffectView.removeEffect(_:)):
(PlatformIntelligenceTextEffectView.removeAllEffects):
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h:
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.swift:
(effectView):
(contextRange):
(activePonderingEffect):
(activeReplacementEffect):
(textVisibilityRegionIdentifiers):
(replacementQueue):
(onFlushCompletion):
(characterDelta(forReceivedSuggestions:)):
(startAnimation(for:)):
(requestReplacement(withProcessedRange:characterDelta:operation:)):
(flushReplacements):
(startReplacementAnimation(using:)):
(setupViewIfNeeded):
(destroyViewIfNeeded):
(setActivePonderingEffect(_:)):
(setActiveReplacementEffect(_:)):
(reset):
(WKIntelligenceTextEffectCoordinator.textPreview(for:)):
(WKIntelligenceTextEffectCoordinator.updateTextChunkVisibility(_:visible:force:)):
(WKIntelligenceTextEffectCoordinator.updateTextChunkVisibility(_:visible:)):
(WKIntelligenceTextEffectCoordinator.performReplacementAndGeneratePreview(for:effect:animation:)):
(WKIntelligenceTextEffectCoordinator.replacementEffectWillBegin(_:)):
(WKIntelligenceTextEffectCoordinator.flushRemainingReplacementsIfNeeded):
(WKIntelligenceTextEffectCoordinator.replacementEffectDidComplete(_:)):
(range):
(WKIntelligenceTextEffectCoordinator.hash(into:)):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::proofreadingSessionDidReceiveSuggestions):
(WebKit::WebPage::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPage::decorateTextReplacementsForActiveWritingToolsSession):
(WebKit::WebPage::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_UpdateTextVisibilityForRange)):

Canonical link: <a href="https://commits.webkit.org/286126@main">https://commits.webkit.org/286126@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a1c8e748cdfa3f29f75fc2783cc4a5ee0172aea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54338 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27725 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79338 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26147 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77025 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/63474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2123 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/17097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77975 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/63474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64365 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/39218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/63474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21864 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/24479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/63474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22208 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80822 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2226 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1347 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2375 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64384 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66385 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16493 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/10320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2191 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/2219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/3140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2226 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -3748,7 +3748,7 @@ InstanceMethodSwizzler swizzler(PAL::getWTWritingToolsClass(), @selector(showToo
     NSRange subrange = NSMakeRange(17, 45); // "I didn't quite here him.\n\nWho's over they're."
 
-    [coordinatorDelegate intelligenceTextEffectCoordinator:coordinator updateTextVisibilityForRange:subrange visible:NO completion:^{
+    [coordinatorDelegate intelligenceTextEffectCoordinator:coordinator updateTextVisibilityForRange:subrange visible:NO identifier:NSUUID.UUID completion:^{
```
fix this

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2187,15 +2192,13 @@ - (void)proofreadingSession:(WTSession *)session didReceiveSuggestions:(NSArray<
-        // FIXME: Wait for the animation to finish before invoking this method.
-        page->proofreadingSessionDidCompletePartialReplacement(*webSession, replacementData, *webContext, finished, [] {
-            // FIXME: Once the animations are implemented, this completion handler will be used to inform the coordinator that this async operation is complete.
+    [_intelligenceTextEffectCoordinator requestReplacementWithProcessedRange:range characterDelta:delta operation:^(void (^completion)(void)) {
```
I think you should use a lambda + `makeBlockPtr` here. At the very least `_page` should be weakly referenced.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2303,10 +2311,26 @@ - (void)intelligenceTextEffectCoordinator:(WKIntelligenceTextEffectCoordinator *
+        // Ensure the rendering of the visibility has actually taken effect.
+        [strongSelf _doAfterNextPresentationUpdate:makeBlockPtr([completion] {
+            completion();
+        }).get()];
```
Couldn't this just be `_doAfterNextPresentationUpdate:completion.get()`?

pxlcoder commented on Source/WebCore/page/IntelligenceTextEffectsSupport.cpp:
```diff
@@ -109,5 +111,29 @@ std::optional<TextIndicatorData> textPreviewDataForRange(Document& document, con
+
+        auto offsetRange = OffsetRange { marker.startOffset(), marker.endOffset() };
+
+        markers.removeMarkers(node, offsetRange, { DocumentMarker::Type::WritingToolsTextSuggestion });
```
Removing markers while iterating over them is concerning?

rr-codes replied:
I thought it was ok as long as they’re being added back in the same iteration, with the same ranges? Or is that still bad?

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -277,12 +291,36 @@ static bool isZeroToOneCompositionType(WritingTools::Session::CompositionType ty
+    //
+    // Because this happens in the same run-loop cycle, this change will not be seen by the user.
+
+    auto replacementLocationOffsetBeforeBatch = static_cast<int>(state->replacementLocationOffset);
```
Why the `static_cast`?

rr-codes replied:
oops, forgot to remove

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -306,68 +344,29 @@ static bool isZeroToOneCompositionType(WritingTools::Session::CompositionType ty
-        ASSERT_NOT_REACHED();
-        return;
-    }
+        auto replacementLocationOffsetAfterBatch = (state->replacementLocationOffset);
```
The parentheses aren't necessary?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2240,9 +2245,19 @@ - (void)didEndWritingToolsSession:(WTSession *)session accepted:(BOOL)accepted
     _activeWritingToolsSession = nil;
     [_writingToolsTextSuggestions removeAllObjects];
 
-    _page->setWritingToolsActive(false);
```
Why is this no longer being called if `session.type != WTSessionTypeProofreading`?

pxlcoder commented on Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:
```diff
@@ -91,23 +91,25 @@ extension PlatformIntelligenceTextEffect {
         }
 
-        return preview
+        fatalError("Failed to create a targeted preview: parameter was of unexpected type \(type(of: chunk)).")
```
If `fatalError` is a release assert, I'm not sure this is safe, since `TextIndicator` creation can fail?

Same question applies to other use of `fatalError` in this patch.

rr-codes replied:
in this particular case, this fatalError is fine, because the only two subclasses of the type are already handled, so it’s impossible for this code to ever get executed

rr-codes replied:
actually, all the fatal errors are safe (there are however a few ! which are not safe, so I’ll fix those)

pxlcoder commented on Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h:
```diff
@@ -34,17 +34,18 @@
 #import <WebKit/_WKTextPreview.h>
 #endif
 
+#import <WebKit/WKWebView.h>
```
Is this important necessary, or can you forward declare here?

rr-codes replied:
it’s needed since Swift does not understand forward declaration types

pxlcoder reviewed (approved):
Just confirming that you ran all the tests?

rr-codes committed:
[Intelligence Effects] Support adding intelligence effects to Writing Tools Proofreading operations
https://bugs.webkit.org/show_bug.cgi?id=282161
rdar://138751347

Reviewed by Aditya Keerthi.

When performing Proofreading, the text will now animate using the intelligence text effects.

The implementation details are described throughout the change in comments.

* Source/WebCore/page/IntelligenceTextEffectsSupport.cpp:
(WebCore::IntelligenceTextEffectsSupport::updateTextVisibility):
(WebCore::IntelligenceTextEffectsSupport::decorateWritingToolsTextReplacements):
* Source/WebCore/page/IntelligenceTextEffectsSupport.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::proofreadingSessionDidReceiveSuggestions):
(WebCore::Page::updateTextVisibilityForActiveWritingToolsSession):
(WebCore::Page::decorateTextReplacementsForActiveWritingToolsSession):
(WebCore::Page::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::didBeginWritingToolsSession):
(WebCore::WritingToolsController::proofreadingSessionDidReceiveSuggestions):
(WebCore::WritingToolsController::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/ObjectiveCBlockConversions.swift:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView proofreadingSession:didReceiveSuggestions:processedRange:inContext:finished:]):
(-[WKWebView didEndWritingToolsSession:accepted:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:identifier:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:decorateReplacementsForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:completion:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::proofreadingSessionDidReceiveSuggestions):
(WebKit::WebPageProxy::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPageProxy::decorateTextReplacementsForActiveWritingToolsSession):
(WebKit::WebPageProxy::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:
(PlatformIntelligenceTextEffectViewSource.replacementEffectDidComplete(_:)):
(UITextEffectViewSourceAdapter.targetedPreview(for:)):
(UITextEffectViewSourceAdapter.updateTextChunkVisibilityForAnimation(_:visible:)):
(UIReplacementTextEffectDelegateAdapter.replacementEffectDidComplete(_:)):
(UIReplacementTextEffectDelegateAdapter.performReplacementAndGeneratePreview(for:effect:animation:)):
(WTTextPreviewAsyncSourceAdapter.textPreviews(for:)):
(WTTextPreviewAsyncSourceAdapter.updateIsTextVisible(_:for:)):
(PlatformIntelligenceTextEffectView.wrappedEffectIDToPlatformEffects):
(PlatformIntelligenceTextEffectView.frame):
(PlatformIntelligenceTextEffectView.removeEffect(_:)):
(PlatformIntelligenceTextEffectView.removeAllEffects):
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h:
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.swift:
(effectView):
(contextRange):
(activePonderingEffect):
(activeReplacementEffect):
(textVisibilityRegionIdentifiers):
(replacementQueue):
(onFlushCompletion):
(characterDelta(forReceivedSuggestions:)):
(startAnimation(for:)):
(requestReplacement(withProcessedRange:characterDelta:operation:)):
(flushReplacements):
(startReplacementAnimation(using:)):
(setupViewIfNeeded):
(destroyViewIfNeeded):
(setActivePonderingEffect(_:)):
(setActiveReplacementEffect(_:)):
(reset):
(WKIntelligenceTextEffectCoordinator.textPreview(for:)):
(WKIntelligenceTextEffectCoordinator.updateTextChunkVisibility(_:visible:force:)):
(WKIntelligenceTextEffectCoordinator.updateTextChunkVisibility(_:visible:)):
(WKIntelligenceTextEffectCoordinator.performReplacementAndGeneratePreview(for:effect:animation:)):
(WKIntelligenceTextEffectCoordinator.replacementEffectWillBegin(_:)):
(WKIntelligenceTextEffectCoordinator.flushRemainingReplacementsIfNeeded):
(WKIntelligenceTextEffectCoordinator.replacementEffectDidComplete(_:)):
(range):
(WKIntelligenceTextEffectCoordinator.hash(into:)):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::proofreadingSessionDidReceiveSuggestions):
(WebKit::WebPage::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPage::decorateTextReplacementsForActiveWritingToolsSession):
(WebKit::WebPage::proofreadingSessionDidCompletePartialReplacement): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_UpdateTextVisibilityForRange)):

Canonical link: https://commits.webkit.org/286126@main
